src/vehicle_move.cpp

```diff
@@ -116,6 +117,261 @@ int vehicle::slowdown( int at_velocity ) const
     return std::max( 1, slowdown );
 }
 
+void vehicle:: smart_controller_handle_turn( bool thrusting )
+{
+
+    if( !engine_on || !has_enabled_smart_controller ) {
+        smart_controller_state = cata::nullopt;
+        return;
+    }
+
+    if( smart_controller_state && smart_controller_state->created == calendar::turn ) {
+        return;
+    }
+
+    // controlled engines
+    // note: contains indices of of elements in `engines` array, not the part ids
+    std::vector<int> c_engines;
+    for( int i = 0; i < static_cast<int>( engines.size() ); ++i ) {
+        if( ( is_engine_type( i, fuel_type_battery ) || is_combustion_engine_type( i ) ) &&
+            ( ( parts[ engines[ i ] ].is_available() && engine_fuel_left( i ) > 0 ) ||
+              is_part_on( engines[ i ] ) ) ) {
+            c_engines.push_back( i );
+        }
+    }
+
+    bool rotorcraft = is_flying && is_rotorcraft();
+
+    if( rotorcraft || c_engines.size() <= 1 || c_engines.size() > 5 ) { // bail and shut down
+        for( const vpart_reference &vp : get_avail_parts( "SMART_ENGINE_CONTROLLER" ) ) {
+            vp.part().enabled = false;
+        }
+
+        if( player_in_control( g->u ) ) {
+            if( rotorcraft ) {
+                add_msg( _( "Smart controller does not support flying vehicles." ) );
+            } else if( c_engines.size() <= 1 ) {
+                add_msg( _( "Smart controller detects only a single controllable engine." ) );
+                add_msg( _( "Smart controller is designed to control more than one engine." ) );
+            } else {
+                add_msg( _( "Smart controller does not support more than five engines." ) );
+            }
+            add_msg( m_bad, _( "Smart controller is shutting down." ) );
+        }
+        has_enabled_smart_controller = false;
+        smart_controller_state = cata::nullopt;
+        return;
+    }
+
+    int cur_battery_level, max_battery_level;
+    std::tie( cur_battery_level, max_battery_level ) = battery_power_level();
+    int battery_level_percent = max_battery_level == 0 ? 0 : cur_battery_level * 100 /
+                                max_battery_level;
+
+    // when battery > 90%, discharge is allowed
+    // otherwise trying to charge battery to 90% within 30 minutes
+    bool discharge_forbidden_soft = battery_level_percent <= 90;
+    bool discharge_forbidden_hard = battery_level_percent <= 25;
+    int target_charging_rate = ( max_battery_level == 0 || !discharge_forbidden_soft ) ? 0 :
+                               ( max_battery_level * 9 / 10 - cur_battery_level ) * 10 / ( 6 * 3 );
```

Comment:
( 9 * max_battery_level - 10 * cur_battery_level ) / ( 6 * 3 )
saves one unnecessary division and shouldn't be any less clear.